doc: releases: Clarify what belongs in relnotes vs migration

carlescufi opened the pull request:
The intention of this commit is to clarify what belongs where, given that recent contributions have made it unclear.

henrikbrixandersen commented on doc/releases/index.rst:
```diff
@@ -74,6 +74,12 @@ so with the help of west.
 The project's technical documentation is also tagged to correspond with a
 specific release and can be found at https://docs.zephyrproject.org/.
 
+Release notes contain a list of changes that have been made to the different
+areas of the project during the development cycle of the release. Those do not
+include changes that require the user to modify their own application in order
+to upgrade to the release. Those instead belong in the corresponding migration
+document.
```
I was hoping for more recommendations on what to put where, whether to link between the two files, etc.

At least a link and an introduction to the split of documentation at the very top of each release note file would be good. I am afraid users/developers will overlook the newly introduced migration guide otherwise.

kartben replied:
Sorry that I, as the assignee, let this one go stale. I agree with @henrikbrixandersen.
Along these lines, I think the current wording implies that any change requiring modifications/migration would *only* be described in the migration guide, which I think would be counterproductive and don't reflect how we'd want to do things.
See e.g commit ac520e6f99ce477f512a797f90ec08b063850fe1: it does have a short entry in the release notes describing something new *and* a related entry in the migration guide as this new features comes with some breaking changes.

Something like this?

```suggestion
Release notes contain a list of changes that have been made to the different
areas of the project during the development cycle of the release. Changes that
require the user to modify their own application to support the new release are 
not to be detailed in the release notes but rather have dedicated entries in the
corresponding migration document
```

mbolivar-ampere reviewed (commented):
@carlescufi 

> require the user to modify their own application

Doesn't this mean that the migration guide should only cover breaking changes, but not deprecations? I think deprecations belong in the migration guide too. As a user, I would certainly like to learn about them earlier rather than later.

I'm trying to sort out the differences for the devicetree API in particular. This is a bit of a special case, as you know. I think you are implying:

- new zephyr/devicetree.h macros go in the release notes
- new bindings go in the release notes
- new properties go in the release notes
- newly deprecated properties go in the release notes (again, this feels a bit anti-user to me)
- the following property changes go in the migration guide: removal, different cells in a specifier, different type, different `const` value, removal of some but not all `enum` elements, new `default` values, going from not required to `required: true`
- removed bindings go in the migration guide

TBH this partition between release notes and migration guide feels artificially fragmented to me.

Wouldn't it make more sense to include all the bindings changes in the release notes, but make special mention of the breaking changes in the migration guide?

henrikbrixandersen commented:
@carlescufi ping

carlescufi commented:
Architecture WG:

- @nashif says that whatever goes to the migration guide must be useful to anyone migrating.
- @henrikbrixandersen added only the new drivers/features in the release notes. In the migration guide the required changes
- @fabiobaltieri agrees to add new features only to the release notes
- @nashif suggests an editorial process at the end that cleans up the contents, removes excess contents and move content to the migration guide
- @kartben mentions that information is missing in some cases, which can be attributed to maintainers not completing the release notes in time

The following belongs in the migration guide and in the migration guide (Required Changes) ONLY :
- Changes in the code that require an action (i.e. a modification of the source code or configuration files) on the part of the user in order to keep the existing behavior of their application. This includes:
   - Breaking API changes
   - Deprecations
   - Devicetree or Kconfig changes that affect the user (changes to defaults, renames, etc)
   - Treewide changes that have an effect (eg. changing the include path or defaulting to a different C standard library)

   - Anything else that can affect the compilation or runtime behavior of an existing application

Remove the sections, and the recommended changes altogether.

fabiobaltieri commented:
Maybe for the release notes we could just have
- Major enhancement (we already have it), single list, prob the same as the blog post
- Other new features by subsystem, pretty much new drivers or smaller features

Then the maintainer would have to keep the "major" list curated, for the subsystem if there's no significant change we'd just ask to list the new drivers, or nothing if there's none (which is probably pretty common).

Then for the template, we were kicking around the idea of adding some new metadata in the maintainers file. Maybe we could have a bit that says if the area should have a section in the release notes and migration notes templates, and then we could use that to generate the template automatically, as well as a checklist like https://github.com/zephyrproject-rtos/zephyr/issues/50423 to ask maintainers to add their section?

mbolivar-ampere commented:
Can someone who was at the Arch WG confirm this impact on DT?

Release notes:

- new zephyr/devicetree.h macros
- new bindings
- new properties

Migration guide:

- removed bindings
- bindings property changes: deprecation, removal and other breaking changes [*]
 
The bindings parts of the release notes are partially automated, so I will update my script accordingly if this is the WG's conclusion.

[*] for the DT stans out there, a complete list is: different cells in a specifier, different type, different const value, removal of some but not all enum elements, new default values, going from not required to required: true

carlescufi commented:
> Can someone who was at the Arch WG confirm this impact on DT?
> 
> Release notes:
> 
> * new zephyr/devicetree.h macros
> * new bindings
> * new properties
> 
> Migration guide:
> 
> * removed bindings
> * bindings property changes: deprecation, removal and other breaking changes [*]

I can confirm this was the intent @mbolivar-ampere

henrikbrixandersen commented:
This needs rebasing. Is the target milestone still v3.6.0?

kartben commented:
@carlescufi could you help get this over the finish line so that we have better (albeit maybe not perfect) guidelines for the upcoming 3.6, as otherwise I fear we'll spend a lot of time repeating ourselves on various PRs to give people guidance

carlescufi commented:
> @carlescufi could you help get this over the finish line so that we have better (albeit maybe not perfect) guidelines for the upcoming 3.6, as otherwise I fear we'll spend a lot of time repeating ourselves on various PRs to give people guidance

@kartben yep, on it! Let's see if I can get it done today.

carlescufi committed:
doc: releases: Clarify what belongs in relnotes vs migration

The intention of this commit is to clarify what belongs where, given
that recent contributions have made it unclear.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@henrikbrixandersen @kartben PTAL

henrikbrixandersen reviewed (dismissed):
Thanks!

kartben commented on doc/releases/index.rst:
```diff
@@ -68,6 +68,12 @@ As of 2022-01-01, LTS1 (1.14.x) is not supported and has reached end of life (EO
+areas of the project during the development cycle of the release.
+Changes that require the user to modify their own application to support the new
+release may be mentioned in the release notes, but the details regarding `what`
+needs to be changed are to be detailed the release's migration guide.
```
sorry, probably my typo there:

```suggestion
needs to be changed are to be detailed in the release's migration guide.
```

kartben commented on doc/releases/index.rst:
```diff
@@ -83,6 +89,20 @@ Migration Guides
+- Deprecations
+- Devicetree or Kconfig changes that affect the user (changes to defaults,
+  renames, etc)
+- Treewide changes that have an effect (eg. changing the include path or
```
```suggestion
- Treewide changes that have an effect (e.g. changing the include path or
```

carlescufi replied:
Fixed.

kartben reviewed (changes requested):
Thank you, sir!
Spotted a typo (probably mine), and a small nit. Thanks!
